framework_tvb/tvb/interfaces/web/controllers/flow_controller.py

```diff
@@ -377,64 +383,40 @@ def accepted__aggregation_functions():
-                    filters[FILTER_VALUES][i] = temp_date
-                except ValueError:
-                    raise
-        # In order for the filter object not to "stack up" on multiple calls to
-        # this method, create a deepCopy to work with
-        if constants.ELEM_CONDITIONS in current_node:
-            new_filter = copy.deepcopy(current_node[constants.ELEM_CONDITIONS])
-        else:
-            new_filter = FilterChain()
-        new_filter.fields.extend(filters[FILTER_FIELDS])
-        new_filter.operations.extend(filters[FILTER_OPERATIONS])
-        new_filter.values.extend(filters[FILTER_VALUES])
-        # Get dataTypes that match the filters from DB then populate with values
-        values, total_count = [], 0
-        # Create a dictionary that matches what the template expects
-        parameters = {ABCAdapter.KEY_NAME: name,
-                      ABCAdapter.KEY_FILTERABLE: availablefilter,
-                      ABCAdapter.KEY_TYPE: constants.TYPE_SELECT,
-                      ABCAdapter.KEY_OPTIONS: values,
-                      ABCAdapter.KEY_DATATYPE: datatype}
-
-        if total_count > MAXIMUM_DATA_TYPES_DISPLAYED:
-            parameters[KEY_WARNING] = WARNING_OVERFLOW
-
-        if constants.ATT_REQUIRED in current_node:
-            parameters[constants.ATT_REQUIRED] = current_node[constants.ATT_REQUIRED]
-            if len(values) > 0 and string2bool(str(parameters[constants.ATT_REQUIRED])):
-                parameters[ABCAdapter.KEY_DEFAULT] = str(values[-1][ABCAdapter.KEY_VALUE])
-        previous_selected = self.context.get_current_default(name)
-        if previous_selected in [str(vv['value']) for vv in values]:
-            parameters[ABCAdapter.KEY_DEFAULT] = previous_selected
-
-        template_specification = {"inputRow": parameters, "disabled": False,
-                                  "parentDivId": parent_div, common.KEY_SESSION_TREE: tree_session_key}
-        return self.fill_default_attributes(template_specification)
+        name_start_index = datatype_index_path.rfind('.')
+        datatype_index_name = datatype_index_path[name_start_index+1:]
+        index_class = getattr(sys.modules[datatype_index_path[: name_start_index]], datatype_index_name)
+        filters_dict = json.loads(filters)
+
+        fields = []
+        operations = []
+        values = []
+
+        for idx in range(len(filters_dict['fields'])):
+            fields.append(filters_dict['fields'][idx])
+            operations.append(filters_dict['operations'][idx])
+            values.append(filters_dict['values'][idx])
+
+        filter = FilterChain(fields=fields, operations=operations, values=values)
+        project = common.get_current_project()
+
+        form = Form(project_id=project.id, draw_ranges=True)
+        data_type_gid_attr = DataTypeGidAttr(linked_datatype=REGISTRY.get_datatype_for_index(index_class()))
+        select_field = TraitDataTypeSelectField(data_type_gid_attr, form, conditions=filter)
```

Comment:
The new logic looks much cleaner that the prev code! Good job!

Not sure if it is really justified here, but could we potentially push some of this logic in a service and leave here in the controller only the part about inputs parsing and link with a jinja/response json part?

Reply:
I think we could, why not... But in which one? OperationService? Or a new service just for filters?

Reply:
flowservice ?

Reply:
I don't think we have a Flow Service, so should I create it?

Reply:
Wasn't it recently removed?

Reply:
`FlowService` was recently renamed into `AlgorithmService` ..... by me ;-)

Reply:
If I read this again, most of the logic is parsing of strings into filters, so it would be right to stay in Controller.